```{r setup}
library(DESeq2)
library(tidyverse)
library(parallel)
library(doParallel)
library(BiocParallel)
library(pheatmap)
library(readxl)
library(uwot)
library(factoextra)
library(biomaRt)
library(ggsci)
library(Cairo)
library(RPushbullet)
library(WGCNA)
library(RColorBrewer)
library(Rsamtools)
library(BSgenome.Mmusculus.UCSC.mm10)
library(BSgenome)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(GenVisR)
library(lubridate)
library(broom)
library(ggsignif)

source([REDACTED_FILE_PATH]
source([REDACTED_FILE_PATH]
source([REDACTED_FILE_PATH]
source([REDACTED_FILE_PATH]
source([REDACTED_FILE_PATH]
source([REDACTED_FILE_PATH]


registerDoParallel(makeCluster(12))
register(MulticoreParam(12))

set.seed(12345)

pal = pal_nejm()(8)
```

```{bash eval=FALSE}
#!/bin/bash
#SBATCH -A [REDACTED_PROJECT_ALLOCATION]
#SBATCH -p genomics
#SBATCH -t 6:00:00
#SBATCH -N 1
#SBATCH --mem=48G
#SBATCH --ntasks-per-node=12
#SBATCH --mail-user=[REDACTED_EMAIL_URL]
#SBATCH --mail-type=BEGIN,END,FAIL
#SBATCH --job-name='[REDACTED_DATE_YYMMDD]mertk_bcl2fastq'

module purge
cd [REDACTED_FILE_PATH]
module load bcl2fastq/2.19.1 #for compatibility with previous expts
bcl2fastq --loading-threads 4 --writing-threads 4 --processing-threads 4
```   

```{bash eval=FALSE}
#!/bin/bash
#SBATCH -A [REDACTED_PROJECT_ALLOCATION]
#SBATCH -p genomics
#SBATCH -t 6:00:00
#SBATCH -N 1
#SBATCH --mem=64G
#SBATCH --ntasks-per-node=18
#SBATCH --mail-user=[REDACTED_EMAIL_URL]
#SBATCH --mail-type=BEGIN,END,FAIL
#SBATCH --job-name='[REDACTED_DATE_YYMMDD]mertk_bclconvert'

module purge
cd [REDACTED_FILE_PATH]
module load bcl-convert/3.10.5 #using new sequencer (Nextseq 2000)
bcl-convert --bcl-input-directory [REDACTED_FILE_PATH] \
--output-directory [REDACTED_FILE_PATH] \
--bcl-num-conversion-threads 6 \
--bcl-num-compression-threads 6 \
--bcl-num-decompression-threads 6
```  

```{r}
#get samples from file names, link to mouse IDs
ss = read_csv([REDACTED_FILE_PATH]
              skip = 15, 
              trim_ws = T)
fastq_files = list.files(path = [REDACTED_FILE_PATH]
                         full.names = T) %>% 
   as_tibble() %>% 
  #originally no column name, so add required one
  dplyr::rename(fastq_1 = value) %>% 
   #only single-end, so leave empty
   mutate(fastq_2 = NA) %>% 
  #location of the sample name in the file path
   dplyr::mutate(Sample_ID = substring(text = fastq_1, 
                                       first = regexpr(pattern = "GRSB", text = fastq_1), 
                                       last = (regexpr(pattern = "S\\d+_L00\\d", text = fastq_1) -  2))) %>% 
   mutate(strandedness = "reverse") %>% 
   #merge with sample sheet using "Sample_ID"
   left_join(., ss) %>% 
   dplyr::select(-contains("index"), -contains("sample", ignore.case = T)) %>% 
   dplyr::rename(mouse = Description)

#link with metadata from mouse IDs
md_mertk = read_excel([REDACTED_FILE_PATH]
                sheet = "Metadata") %>% 
  dplyr::rename(mouse = full_name) %>% 
  right_join(., fastq_files) %>% 
  mutate(Group = ifelse(mouse == "UmRNA", yes = "uRNA", no = Group)) %>% 
  #to separate between batches
  dplyr::mutate(mouse = ifelse(mouse == "UmRNA",
                               yes = "uRNA_Mertk_batch1",
                               no = mouse),
                batch = "Mertk_batch1",
                Group = ifelse(Group == "Cx3cr1-CreER; zsGreen x Mertk flox",
                                  yes = "Cx3cr1-CreER; zsGreen x Mertk flox, 16 DPT",
                                  no = Group)) %>% 
  dplyr::select(-c(Mouse, Cage)) %>% 
  dplyr::rename(sample = mouse) %>% 
  dplyr::relocate(sample, fastq_1, fastq_2, strandedness) 
```   

```{r}
#get samples from file names, link to mouse IDs
ss = read_csv([REDACTED_FILE_PATH]
              skip = 15, 
              trim_ws = T)
fastq_files = list.files(path = [REDACTED_FILE_PATH]
                         pattern = "^AM.+\\.fastq\\.gz",
                         recursive = T,
                         full.names = T) %>% 
   as_tibble() %>% 
  #originally no column name, so add required one
  dplyr::rename(fastq_1 = value) %>% 
   #only single-end, so leave empty
   mutate(fastq_2 = NA) %>% 
  #location of the sample name in the file path
   dplyr::mutate(Sample_ID = substring(text = fastq_1, 
                                       first = regexpr(pattern = "AM_21|AM_NC_21", text = fastq_1), 
                                       last = (regexpr(pattern = "S\\d+_L00\\d", text = fastq_1) -  2))) %>% 
   mutate(strandedness = "reverse") %>% 
   #merge with sample sheet using "Sample_ID"
   left_join(., ss) %>% 
   dplyr::select(-contains("index"), -contains("sample", ignore.case = T)) %>% 
   dplyr::rename(mouse = Description)

#link with metadata from mouse IDs
md_controls = read_excel([REDACTED_FILE_PATH]
                sheet = "Group Assignments",
                range = "A1:Q79") %>% 
   mutate(mouse = paste(substring(cage, nchar(cage) - 3), ear_tag, sep = "_")) %>% 
   right_join(., fastq_files) %>% 
  mutate(Group = ifelse(mouse == "uRNA", yes = "uRNA", no = "Wild-Type Controls, Naïve")) %>% 
  #to separate between batches
  dplyr::mutate(mouse = ifelse(mouse == "uRNA",
                               yes = "uRNA_2020_2021_controls",
                               no = mouse),
                batch = "2020_2021_controls") %>% 
  #select naïve
  dplyr::filter((IAV == "Vehicle" & ISRIB == "Vehicle") | mouse ==  "uRNA_2020_2021_controls") %>% 
  dplyr::mutate(IAV = FALSE,
                tamoxifen = FALSE,
                wheels = F) %>% 
  dplyr::rename(sample = mouse) %>% 
  dplyr::select(sample, fastq_1, fastq_2, strandedness, age_group, Group, genotype,
                takedown_date, birthdate, IAV, tamoxifen, batch, wheels) %>% 
  dplyr::relocate(sample, fastq_1, fastq_2, strandedness) 
```   

```{r}
#get samples from file names, link to mouse IDs
ss = read_csv([REDACTED_FILE_PATH]
              skip = 41, 
              trim_ws = T)
fastq_files = list.files(path = [REDACTED_FILE_PATH]
                         pattern = "^\\d{4}_.+\\.fastq\\.gz|[REDACTED_IDENTIFIER]_urna2ng_S9_R1_001.fastq.gz",
                         recursive = T,
                         full.names = T) %>% 
   as_tibble() %>% 
  #originally no column name, so add required one
  dplyr::rename(fastq_1 = value) %>% 
   #only single-end, so leave empty
   mutate(fastq_2 = NA) %>% 
  #location of the sample name in the file path
   dplyr::mutate(Sample_ID = substring(text = fastq_1, 
                                       first = regexpr(pattern = "\\d{4}_\\d_S\\d", text = fastq_1), 
                                       last = (regexpr(pattern = "S\\d+_R1", text = fastq_1) -  2))) %>% 
  dplyr::mutate(Sample_ID = ifelse(grepl("urna", Sample_ID),
                                   yes = "uRNA_mertkflox_60DPT",
                                   no = Sample_ID)) %>% 
   mutate(strandedness = "reverse") %>% 
   #merge with sample sheet using "Sample_ID"
   left_join(., ss) %>% 
   dplyr::select(-c(ProjectName:ncol(.))) %>% 
   dplyr::rename(mouse = Sample_ID) 

#link with metadata from mouse IDs
md_mertk_60dpt = read_excel([REDACTED_FILE_PATH]
                sheet = "Metadata") %>% 
   mutate(mouse = paste(substring(Cage, nchar(Cage) - 3), Mouse, sep = "_")) %>% 
   right_join(., fastq_files) %>% 
  #to separate between batches
  dplyr::mutate(batch = "Mertk-Flox_60DPT") %>% 
  dplyr::mutate(IAV = FALSE,
                tamoxifen = FALSE,
                wheels = F,
                Group = case_when(grepl("uRNA", mouse) ~ "uRNA",
                                  Group == "Mertk-flox 60DPT" ~ "Cx3cr1-CreER; zsGreen x Mertk flox, 61 DPT",
                                  Group == "Control" ~ "Cx3cr1-CreER x zsGreen Control",
                                  TRUE ~ Group)) %>% 
  dplyr::rename(sample = mouse) %>% 
  dplyr::select(sample, fastq_1, fastq_2, strandedness, age_group, Group, genotype,
                takedown_date, birthdate, IAV, tamoxifen, batch, wheels, days_after_tamoxifen) %>% 
  dplyr::relocate(sample, fastq_1, fastq_2, strandedness) 
```   

```{r}
md = bind_rows(md_mertk, md_controls, md_mertk_60dpt) %>% 
  dplyr::mutate(Group = factor(Group),
                IAV = factor(IAV),
                tamoxifen = factor(tamoxifen),
                wheels = factor(wheels),
                takedown_date = as.Date(takedown_date),
                birthdate = as.Date(birthdate),
                age_at_takedown = interval(birthdate, takedown_date) %/% months(1),
                genotype = gsub("zsGreen$", "zsGreen-flox", genotype),
                genotype = gsub("zsGreen x", "zsGreen-flox x", genotype),
                genotype = gsub("Mertk flox", "Mertk-flox", genotype),
                Mertk = ifelse(genotype %in% c("Cx3cr1-CreER x zsGreen", "C57BL/6J"),
                                      yes = "Wild-Type",
                                      no = genotype),
                Mertk = factor(Mertk),
                genotype = factor(genotype))

write_csv(md,
          [REDACTED_FILE_PATH]
          na = "")
```

```{bash eval=FALSE}
screen
cd [REDACTED_FILE_PATH]
module purge
module load singularity
module load graphviz/2.40.1


nextflow run nf-core/rnaseq \
-r '3.5' \
-profile nu_genomics \
--email '[REDACTED_EMAIL_URL]' \
--additional_fasta [REDACTED_FILE_PATH] \
--star_index false \
--three_prime_clip_r2 3 \
--genome 'GRCm38' \
-work-dir [REDACTED_FILE_PATH] \
--input [REDACTED_FILE_PATH]
```

```{r}
#import gene conversion
mouse_conv = read_tsv([REDACTED_FILE_PATH] %>% 
   dplyr::select(ensembl_gene_id = gene_id, external_gene_name = gene_name)

#import des
load([REDACTED_FILE_PATH] 
dds = dds[, !grepl("uRNA|UmRNA", dds$sample)] #remove URNAs, which we know from multiQC cluster out, close together
tmp = colData(dds) %>% 
   as.data.frame() %>% 
   mutate(mouse = substring(text = sample, first = 2)) %>% 
   dplyr::select(-sizeFactor)
md_final = md %>% 
   dplyr::select(-contains("fastq"), -strandedness) %>% 
   unique()
md_final = left_join(tmp, md_final, by = c("mouse" = "sample")) %>% 
   column_to_rownames("sample") 
all(colnames(dds) == rownames(md_final)) #TRUE
des = DESeqDataSetFromMatrix(countData = counts(dds, normalized = F), 
                                  colData = md_final,
                                  design = ~ Group)
```   

```{r}
ggplot(subset(md_final, Group != "uRNA"), aes(x = Group, y = age_at_takedown, fill = age_group)) +
   geom_boxplot(outlier.shape = NA) +
   geom_point(position = position_jitterdodge(jitter.width = 0.2, seed = 42)) +
   scale_fill_nejm(name = "Age Group") +
   theme_bw() +
   theme(strip.background = element_blank(),
        axis.text.x = element_text(angle = 45, hjust=1)) +
  labs(x = "Experimental Group", y = "Age at Takedown (months)") +
  ylim(0, NA)
```

```{r}
pca = plotPCA_manual(object = vst(des), 
                     intgroup = "Group", 
                     pcs = 3, 
                     ntop = nrow(des),
                     merge_metadata = TRUE,
                     return_loadings = TRUE,
                     custom_annotation = mouse_conv)

ggplot(pca$data, aes(x = PC1, y = PC2, color = Group, shape = age_group)) +
   geom_point(size = 3)  +
   labs(x = paste0("PC1 (", pca$percent_var$percent_var[1], "% of variance explained)"),
        y = paste0("PC2 (", pca$percent_var$percent_var[2], "% of variance explained)")) +
   theme_bw(base_family = "Arial") +
   scale_color_nejm()
ggplot(pca$data, aes(x = PC2, y = PC3, color = Group, shape = age_group)) +
   geom_point(size = 3) +
   labs(x = paste0("PC2 (", pca$percent_var$percent_var[2], "% of variance explained)"),
        y = paste0("PC3 (", pca$percent_var$percent_var[3], "% of variance explained)")) +
   theme_bw(base_family = "Arial") +
   scale_color_nejm()
```  

```{r}
parametric = DESeq(des,
                   fitType = "parametric",
                   parallel = T)
plotDispEsts(parametric) #realistically this is pretty good

local = DESeq(des,
              fitType = "local",
              parallel = T)
plotDispEsts(local) #modestly better, keep

dge = local
rm(local, parametric)
```

```{r}
umap_counts = vst(dge, blind = F, fitType = "local") %>% 
  assay() %>% 
   t()
all_pca = prcomp(umap_counts)
fviz_eig(all_pca, 
         barfill = "dodgerblue4", 
         xlab = "Principal Component",
         ylab = "Percent Variance Explained", 
         main = "",
         ncp = 50,
         ggtheme = theme_bw(),
         addlabels = T)
```   

```{r}
set.seed(12345)
umap_results = umap(umap_counts,
                    pca = 6,
                    pca_center = T,
                    n_threads = 1,
                    scale = "Z",
                    min_dist = 0.01)
rownames(umap_results) = rownames(umap_counts)
colnames(umap_results) = c("UMAP1", "UMAP2")
umap_results = umap_results %>% 
   as_tibble(rownames = NA) %>% 
   mutate(mouse = substring(rownames(.), 2)) %>% 
   left_join(., md_final) 
```

```{r}
ggplot(umap_results, aes(x = UMAP1, y = UMAP2, color = Group, shape = age_group)) +
   geom_point(size = 3) +
   theme_bw(base_family = "Arial") +
   scale_color_nejm()
```   

```{r}
#remove uninteresting for talks
dge_naive = dge[, dge$IAV != TRUE & dge$Group != "Cx3cr1-CreER; zsGreen x Mertk flox, 16 DPT, Young"]
dge_naive$mertk_age = factor(paste(dge_naive$age_group, dge_naive$Mertk, sep = ", "),
                                   levels = c("Young, Wild-Type", "Young, Cx3cr1-CreER x zsGreen-flox",
                                              "Young, Cx3cr1-CreER; zsGreen-flox x Mertk-flox", "Young, Mertk-/-",
                                              "Old, Wild-Type", "Old, MertkCR"))
design(dge_naive) = ~ mertk_age

#also remove transgenes from clustering
not_transgenes = setdiff(rownames(dge), c("CreERT2_gene", "ZsGreen_AF168422.1_gene"))
umap_counts_naive = vst(dge_naive[not_transgenes, ], 
                        blind = F, fitType = "local") %>% 
  assay() %>% 
   t()
all_pca_naive = prcomp(umap_counts_naive)
fviz_eig(all_pca_naive, 
         barfill = "dodgerblue4", 
         xlab = "Principal Component",
         ylab = "Percent Variance Explained", 
         main = "",
         ncp = 50,
         ggtheme = theme_bw(),
         addlabels = T)
```   

```{r}
set.seed(12345)
umap_results_naive = umap(umap_counts_naive,
                    pca = 6,
                    pca_center = T,
                    n_threads = 1,
                    scale = "Z",
                    min_dist = 0.01)
rownames(umap_results_naive) = rownames(umap_counts_naive)
colnames(umap_results_naive) = c("UMAP1", "UMAP2")
umap_results_naive = umap_results_naive %>% 
   as_tibble(rownames = NA) %>% 
   mutate(mouse = substring(rownames(.), 2)) %>% 
   left_join(., md_final) %>% 
  mutate(mertk_age = factor(paste(dge_naive$age_group, dge_naive$Mertk, sep = ", "),
                                   levels = c("Young, Wild-Type", "Young, Cx3cr1-CreER x zsGreen-flox",
                                              "Young, Cx3cr1-CreER; zsGreen-flox x Mertk-flox", "Young, Mertk-/-",
                                              "Old, Wild-Type", "Old, MertkCR")))
```

```{r}
naive_umap = ggplot(umap_results_naive, aes(x = UMAP1, y = UMAP2, color = mertk_age, shape = age_group)) +
   geom_point(size = 3) +
   theme_bw(base_family = "Arial") +
   scale_color_npg(name = "Genotype") +
  scale_shape_discrete(name = "Group") +
  ggforce::geom_mark_ellipse(aes(color = mertk_age, label = mertk_age), 
                             linetype = 2, label.family = "Arial", 
                             label.fill = NA, label.fontface = "plain",
                             con.size = 0.2, show.legend = F, label.fontsize = 10)

CairoPDF([REDACTED_FILE_PATH] 
         width = 12,
         height = 9)
naive_umap
dev.off()

CairoPNG([REDACTED_FILE_PATH] 
         width = 12,
         height = 9, 
         res = 300,
         units = "in")
naive_umap
dev.off()

naive_umap
```   

```{r}
transgenes = get_tidy_counts(obj = dge, 
                        goi = c("CreERT2_gene", "ZsGreen_AF168422.1_gene", "Mertk"), 
                        goi_format = "external_gene_name", 
                        custom_annotation = mouse_conv)
ggplot(transgenes, aes(x = Group, y = counts, fill = Group, shape = age_group)) +
   facet_wrap(~external_gene_name, scales = "free_y") +
   geom_boxplot(outlier.shape = NA) +
   geom_point(position = position_jitterdodge(jitter.width = 0.2, seed = 12345)) +
   scale_fill_nejm() +
   theme_bw() +
   theme(strip.background = element_blank(),
        axis.text.x = element_text(angle = 75, hjust=1)) +
  labs(x = "", y = "Gene Counts")
```

```{r}
mertk_start = [REDACTED_IDENTIFIER]
mertk_end = [REDACTED_IDENTIFIER]

mertk_seq = seq(from = mertk_start, to = mertk_end, by = 1)
mertk_bam = data.frame(chrom = rep(2, (length(mertk_seq) - 1)),
                       start = mertk_seq[1:(length(mertk_seq) - 1)], 
                       end = mertk_seq[2:(length(mertk_seq))],
                       name = paste("Mertk", 1:(length(mertk_seq) - 1), sep = "_"),
                       score = rep(0, (length(mertk_seq) - 1)),
                       strand = rep("+", (length(mertk_seq) - 1)))

write.table(mertk_bam, 
            [REDACTED_FILE_PATH]
            row.names = F,
            col.names = F,
            sep = "\t",
            quote = F)
```

```{bash eval=FALSE}
module load bedtools/2.29.2
cd [REDACTED_FILE_PATH]
bams=[REDACTED_FILE_PATH]
echo $bams > mertk_binned_coverage_order.txt

bedtools multicov -bams $bams -bed mertk_binned.bed -S > mertk_binned_coverage.tsv
```   

```{bash eval=FALSE}
#!/bin/bash
#SBATCH -A [REDACTED_PROJECT_ALLOCATION]
#SBATCH -p genomics
#SBATCH -t 4:00:00
#SBATCH -N 1
#SBATCH --mem=32G
#SBATCH --ntasks-per-node=1
#SBATCH --mail-user=[REDACTED_EMAIL_URL]
#SBATCH --mail-type=BEGIN,END,FAIL
#SBATCH --job-name='[REDACTED_DATE_YYMMDD]mertk_coverage_complete'

module load bedtools/2.29.2
cd [REDACTED_FILE_PATH]
bams=[REDACTED_FILE_PATH]
echo $bams > mertk_binned_coverage_order.txt

bedtools multicov -bams $bams -bed mertk_binned.bed -S > mertk_binned_coverage.tsv
```

```{r}
bams = readLines([REDACTED_FILE_PATH] %>% 
  str_split(" ") %>% 
  unlist() %>% 
  basename() %>% 
  gsub(".markdup.sorted.bam", "", .)

coverage = read.delim([REDACTED_FILE_PATH] 
                      header = F, 
                      col.names = c("chrom", "start", "end", "name", "score", "strand", bams)) %>% 
   dplyr::mutate(chrom = "chr2") %>%  #to fit UCSC nomenclature
   pivot_longer(cols = 7:ncol(.),
                names_to = "sample",
                values_to = "cov") %>% 
   dplyr::mutate(sample = substring(sample, 2))

coverage_list = lapply(bams, function(s){
   sub = coverage %>% 
      dplyr::filter(sample == s)
   
   return(sub) })
names(coverage_list) = bams
```

```{r eval=FALSE}
mm10 = BSgenome.Mmusculus.UCSC.mm10
tx_db = TxDb.Mmusculus.UCSC.mm10.knownGene
gr = GRanges(seqnames=c("chr2"), ranges=IRanges(start = mertk_start, end=mertk_end))
```

```{r eval=FALSE}
cairo_pdf([REDACTED_FILE_PATH] 
         width = 30,
         height = 100)
coverage_plot = genCov(x = coverage_list, 
       txdb = tx_db, 
       gr = gr, 
       genome = mm10, 
       #reduce = T,
       gene_isoformSel = "ENSMUST[REDACTED_IDENTIFIER].4",
       cov_plotType="bar")
dev.off()
```

```{r}
complete_elbow = k_elbow(dge, design = "~ Group", cores = 12)
complete_elbow
```   

```{r eval=FALSE}
kmeans_complete = k_means_figure(dge = dge,
                                 design = "~ Group",
                                 fitType = "local",
                                 k = 6,
                                 go_annotations = "org.Mm.eg.db",
                                 ensembl_db = "mmusculus_gene_ensembl",
                                 legend_factors = c("IAV", "age_group", "Mertk", "days_after_tamoxifen"),
                                 breaks = seq(-3, 3, length.out=101),
                                 cores = 1,
                                 return_genes = T,
                                 display_go_terms = F,
                                 return_go_terms = T,
                                 cluster_columns = T,
                                 show_rownames = F,
                                 qval_cutoff = 0.05,
                                 tidy_go = T)

saveRDS(kmeans_complete, [REDACTED_FILE_PATH]
write.csv(kmeans_complete$genes,  [REDACTED_FILE_PATH]
write.csv(kmeans_complete$GO,  [REDACTED_FILE_PATH]

pbPost(title = "k means complete")
```   

```{r}
kmeans = readRDS([REDACTED_FILE_PATH]
kmeans$plot

CairoPDF([REDACTED_FILE_PATH] 
         width = 8,
         height = 6)
kmeans$plot
dev.off()

CairoPNG([REDACTED_FILE_PATH] 
         width = 8,
         height = 6, 
         res = 300,
         units = "in")
kmeans$plot
dev.off()
```

```{r warning=FALSE}
dam_genes = read_excel([REDACTED_FILE_PATH]
                       skip = 1,
                       col_names = c("external_gene_name",
                                     "UMI_per_cell_homeostatic",
                                     "Log2FC_DAM_over_homeostatic",
                                     "P_p",
                                     "P_fdr"), 
                       col_types = c("text",
                                     "numeric",
                                     "numeric",
                                     "numeric",
                                     "numeric")) %>% 
  #calculated their FDR values wrong? ~Double what I'm seeing, but tightly correlated
  dplyr::mutate(pval = 10 ^ -P_p,
                fdr_original = 10 ^ -P_fdr,
                padj = p.adjust(pval, method = "fdr"),
                UMI_DAM = UMI_per_cell_homeostatic * 2 ^ Log2FC_DAM_over_homeostatic) %>% 
  #only want positive hits for this purpose, and pick genes with reasonable expression
  dplyr::filter(padj < 0.05 & 
                  Log2FC_DAM_over_homeostatic > 0 & 
                  UMI_DAM >= 0.5 & 
                  !grepl("^Rpl|^Rps|^Mt-", external_gene_name)) %>% 
  inner_join(., mouse_conv) %>% 
  dplyr::mutate(gene_list = "DAM Markers\n(Keren-Schaul et al., 2017)")

nrow(dam_genes) #169
```   

```{r}
dge$Group = factor(paste(dge$Group, dge$age_group, sep = ", "))

md = colData(dge) %>% 
  as_tibble(rownames = NA)

dam_counts = counts(dge, normalized = T) %>% 
  as_tibble(rownames = NA) %>% 
  rownames_to_column("ensembl_gene_id") %>% 
  left_join(mouse_conv, na_matches = "never") %>% 
  right_join(dam_genes, na_matches = "never") %>% 
  pivot_longer(cols = starts_with("X"),
               names_to = "mouse",
               values_to = "counts") %>% 
  dplyr::mutate(mouse = substring(mouse, 2)) %>% 
  left_join(., md)

#now sum by sample
dam_sums = dam_counts %>% 
  group_by(mouse, Group, age_group, Mertk, IAV) %>% 
  dplyr::summarize(dam_sum = sum(counts, na.rm = T))
```

```{r}
ggplot(dam_sums, aes(x = Group, y = dam_sum, fill = Group)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 60, hjust=1)) +
  scale_fill_npg() +
  labs(x = "", y = "Summed DAM Gene Expression")
```

```{r}
dam_sums_naive = dam_sums %>% 
  dplyr::filter(IAV == FALSE &
                  #16dpi was too short, just initial try
                  Group != "Cx3cr1-CreER; zsGreen x Mertk flox, 16 DPT, Young") %>% 
  dplyr::mutate(mertk_age = factor(paste(age_group, Mertk, sep = ", "),
                                   levels = c("Young, Wild-Type", "Young, Cx3cr1-CreER x zsGreen-flox",
                                              "Young, Cx3cr1-CreER; zsGreen-flox x Mertk-flox", "Young, Mertk-/-",
                                              "Old, Wild-Type", "Old, MertkCR")))

dam_comps_naive = pairwise.wilcox.test(x = dam_sums_naive$dam_sum, g = dam_sums_naive$mertk_age, 
                                       p.adjust.method = "fdr") %>% 
  tidy() %>% 
  dplyr::filter(p.value < 0.05) %>% 
  dplyr::mutate(yval = seq(from = max(dam_sums_naive$dam_sum), 
                           by = max(dam_sums_naive$dam_sum) * 0.02, 
                           length.out = n()),
                annot = format(p.value, scientific = T, digits = 2))
```

```{r}
dam_plot_naive = ggplot(dam_sums_naive, aes(x = mertk_age, y = dam_sum, fill = mertk_age)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, hjust=1)) +
  scale_fill_npg() +
  labs(x = "", y = "Summed DAM Gene Expression") +
  geom_signif(inherit.aes = F, 
              data = dam_comps_naive,
              aes(xmin = group1, xmax = group2, annotations = annot, y_position = yval),
              tip_length = 0,
              textsize = 3, 
              manual=TRUE) 

CairoPDF([REDACTED_FILE_PATH]
    width = 9,
    height = 8,
    family = "Arial")
dam_plot_naive
dev.off()
CairoPNG([REDACTED_FILE_PATH]
    width = 9,
    height = 8,
    units = "in",
    res = 300,
    family = "Arial")
dam_plot_naive
dev.off()

dam_plot_naive
```

```{r}
dam_markers = c("Cx3cr1", "Cd9", "Hexb", "Apoe", "Trem2", "Cst7", "Hexb")
dam_marker_counts = get_tidy_counts(dge, goi = dam_markers, custom_annotation = mouse_conv, goi_format = "external_gene_name")

ggplot(dam_marker_counts, aes(x = Group, y = counts, fill = Group)) +
  facet_wrap(~external_gene_name, scales = "free_y") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 75, hjust=1)) +
  scale_fill_npg() +
  labs(x = "", y = "Counts")
```

```{r}
minimal = dge[not_transgenes, dge$Group %in% c("Cx3cr1-CreER x zsGreen Control, Young", 
                                               "Cx3cr1-CreER; zsGreen x Mertk flox, 61 DPT, Young", 
                                               "MertkCR, Naïve, Old",
                                               "Wild-Type Controls, Naïve, Old",
                                               "Wild-Type Controls, Naïve, Young")]
#fix awkward naming
minimal$Mertk = as.character(minimal$Mertk)
minimal$Mertk = gsub("zsGreen-flox", "zsGreen-lsl", minimal$Mertk)
minimal$Mertk = gsub("flox", "flox-flox", minimal$Mertk)
minimal$Mertk = factor(minimal$Mertk)
minimal$mertk_age = factor(paste(minimal$age_group, minimal$Mertk, sep = ", "),
                                   levels = c("Young, Wild-Type", "Young, Cx3cr1-CreER x zsGreen-lsl",
                                              "Young, Cx3cr1-CreER; zsGreen-lsl x Mertk-flox-flox",
                                              "Old, Wild-Type", "Old, MertkCR"))
                           
design(minimal) = ~ mertk_age
minimal = DESeq(minimal, fitType = "local", parallel = T)
```   

```{r}
pca_minimal = plotPCA_manual(object = vst(minimal, blind = F, fitType = "local"), 
                     intgroup = "Group", 
                     pcs = 3, 
                     ntop = nrow(des),
                     merge_metadata = TRUE,
                     return_loadings = TRUE,
                     custom_annotation = mouse_conv)

pc12_minimal = ggplot(pca_minimal$data, aes(x = PC1, y = PC2, color = Mertk, shape = age_group)) +
   geom_point(size = 3)  +
   labs(x = paste0("PC1 (", pca_minimal$percent_var$percent_var[1], "% of variance explained)"),
        y = paste0("PC2 (", pca_minimal$percent_var$percent_var[2], "% of variance explained)")) +
   theme_bw(base_family = "Arial") +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        axis.title.x = element_text(size = 18),
        axis.title.y = element_text(size = 18)) +
   scale_color_npg()
pc12_minimal 

ggplot(pca_minimal$data, aes(x = PC2, y = PC3, color = Mertk, shape = age_group)) +
   geom_point(size = 3) +
   labs(x = paste0("PC2 (", pca_minimal$percent_var$percent_var[2], "% of variance explained)"),
        y = paste0("PC3 (", pca_minimal$percent_var$percent_var[3], "% of variance explained)")) +
   theme_bw(base_family = "Arial") +
   scale_color_nejm()

CairoPDF([REDACTED_FILE_PATH]
    width = 9,
    height = 6,
    family = "Arial")
pc12_minimal
dev.off()
CairoPNG([REDACTED_FILE_PATH]
    width = 9,
    height = 6,
    units = "in",
    res = 300,
    family = "Arial")
pc12_minimal
dev.off()

saveRDS(pc12_minimal, [REDACTED_FILE_PATH]
```   

```{r}
pairwise_minimal = get_pairwise_DESeq(minimal, comparison_col = "mertk_age", fit_type = "local", 
                                      sort_direction = "factor_order", df_output = T, 
                                      save_csv = T, cores = 12, 
                                      output_directory = [REDACTED_FILE_PATH]
                                      custom_annotation = mouse_conv)
```

```{r}
mouse_hallmark = fgsea::gmtPathways([REDACTED_FILE_PATH]
ifna_response = data.frame(external_gene_name = mouse_hallmark$HALLMARK_INTERFERON_ALPHA_RESPONSE) %>% 
  left_join(., mouse_conv) %>% 
  dplyr::mutate(gene_list = "Hallmark Interferon Alpha Response\n(MM3877)")

ifng_response = data.frame(external_gene_name = mouse_hallmark$HALLMARK_INTERFERON_GAMMA_RESPONSE) %>% 
  left_join(., mouse_conv) %>% 
  dplyr::mutate(gene_list = "Hallmark Interferon Gamma Response\n(MM3878)")

modules = bind_rows(dam_genes, ifna_response, ifng_response) %>% 
  dplyr::mutate(gene_list = factor(gene_list)) %>% 
  dplyr::filter(!is.na(ensembl_gene_id)) %>% 
  dplyr::select(ensembl_gene_id, external_gene_name, gene_list)
```

```{r}
module_counts = get_tidy_counts(minimal, 
                                goi = unique(modules$ensembl_gene_id), 
                                goi_format = "ensembl_gene_id", 
                                custom_annotation = mouse_conv) %>% 
   dplyr::mutate(mertk_age = paste(age_group, Mertk, sep = ", "),
                 #simplify labels for easier reading
                 mertk_age = gsub("Wild-Type", "WT", as.character(mertk_age)),
                 mertk_age = gsub("Cx3cr1-CreER x zsGreen-lsl", "Driver Control", mertk_age), 
                 mertk_age = gsub("Cx3cr1-CreER; zsGreen-lsl x Mertk-flox-flox", "Mertk KO", mertk_age),
                 mertk_age = factor(mertk_age, levels = c("Young, WT", "Young, Driver Control",
                                                          "Young, Mertk KO",
                                                          "Old, WT", "Old, MertkCR"))) %>% 
  left_join(., modules) %>% 
  group_by(mouse, mertk_age, age_group, gene_list) %>% 
  dplyr::summarize(summed_counts = sum(counts, na.rm = T)) %>% 
  ungroup()
```

```{r}
module_comps = lapply(unique(module_counts$gene_list), function(list){
  sub = subset(module_counts, gene_list == list)
  comps = pairwise.t.test(x = sub$summed_counts, g = sub$mertk_age, 
                                       p.adjust.method = "none") %>%
    tidy() %>% 
    dplyr::mutate(gene_list = list,
                  max_y = max(sub$summed_counts)) }) %>% 
  bind_rows() %>% 
  dplyr::mutate(padj = p.adjust(p.value, method = "fdr")) %>% 
  dplyr::filter(padj < 0.05) %>% 
  group_by(gene_list) %>% 
  dplyr::mutate(yval = seq(from = first(max_y), 
                           by = first(max_y) * 0.03, 
                           length.out = n()),
                annot = format(padj, scientific = T, digits = 2)) %>% 
  ungroup() %>% 
  dplyr::mutate(group1 = factor(group1, levels = c("Young, WT", "Young, Driver Control",
                                                     "Young, Mertk KO",
                                                     "Old, WT", "Old, MertkCR")),
                group2 = factor(group2, levels = c("Young, WT", "Young, Driver Control",
                                                     "Young, Mertk KO",
                                                     "Old, WT", "Old, MertkCR")))
```

```{r}
module_plot = ggplot(module_counts, aes(x = mertk_age, y = summed_counts, fill = mertk_age)) +
  facet_wrap(~ gene_list, scales = "free_y") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        strip.text.x = element_text(size = 16),
        axis.title.y = element_text(size = 18)) +
  scale_fill_npg() +
  labs(x = "", y = "Summed Module Gene Expression") +
  geom_signif(inherit.aes = F, 
              data = module_comps,
              aes(xmin = group1, xmax = group2, annotations = annot, y_position = yval),
              tip_length = 0,
              textsize = 5, 
              manual=TRUE) 

CairoPDF([REDACTED_FILE_PATH]
    width = 12,
    height = 8,
    family = "Arial")
module_plot
dev.off()
CairoPNG([REDACTED_FILE_PATH]
    width = 12,
    height = 8,
    units = "in",
    res = 300,
    family = "Arial")
module_plot
dev.off()

saveRDS(module_plot, [REDACTED_FILE_PATH]

module_plot
```

```{r}
aging_markers = c("Cd9", "Apoe", "Lpl","Cst7", "Itgax", "Spp1", "Ccl12", "Ifit2", "Ifit3", "Ifi204", "Oasl2", "Trim30a")
aging_marker_counts = get_tidy_counts(minimal, goi = aging_markers, custom_annotation = mouse_conv, goi_format = "external_gene_name") %>% 
  dplyr::mutate(external_gene_name = factor(external_gene_name,
                                            levels = c("Apoe", "Cd9", "Cst7", "Itgax", "Lpl", "Spp1",
                                                        "Ccl12", "Ifit2", "Ifit3", "Ifi204", "Oasl2", "Trim30a")),
                #simplify labels for easier reading
                mertk_age = gsub("Wild-Type", "WT", as.character(mertk_age)),
                mertk_age = gsub("Cx3cr1-CreER x zsGreen-lsl", "Driver Control", mertk_age), 
                mertk_age = gsub("Cx3cr1-CreER; zsGreen-lsl x Mertk-flox-flox", "Mertk KO", mertk_age),
                mertk_age = factor(mertk_age, levels = c("Young, WT", "Young, Driver Control",
                                              "Young, Mertk KO",
                                              "Old, WT", "Old, MertkCR")))

#get comparisons from pairwise DESeq
max_counts = aging_marker_counts %>% 
  group_by(external_gene_name) %>% 
  dplyr::summarize(max_counts = max(counts)) %>% 
  ungroup()
#join together all hits
pairwise_comps = lapply(names(pairwise_minimal$hits), function(comp){
  comp_split = str_split(comp, pattern = "_over_") %>% 
    unlist()
  out = pairwise_minimal$hits[[comp]] %>% 
    dplyr::mutate(group1 = gsub("mertk_age_", "", comp_split[1]),
                  group2 = comp_split[2],
                  #simplify labels for easier reading
                  group1 = gsub("Wild-Type", "WT", as.character(group1)),
                  group1 = gsub("Cx3cr1-CreER x zsGreen-lsl", "Driver Control", group1), 
                  group1 = gsub("Cx3cr1-CreER; zsGreen-lsl x Mertk-flox-flox", "Mertk KO", group1),
                  group1 = factor(group1, levels = c("Young, WT", "Young, Driver Control",
                                                     "Young, Mertk KO",
                                                     "Old, WT", "Old, MertkCR")),
                  group2 = gsub("Wild-Type", "WT", as.character(group2)),
                  group2 = gsub("Cx3cr1-CreER x zsGreen-lsl", "Driver Control", group2), 
                  group2 = gsub("Cx3cr1-CreER; zsGreen-lsl x Mertk-flox-flox", "Mertk KO", group2),
                  group2 = factor(group2, levels = c("Young, WT", "Young, Driver Control",
                                                     "Young, Mertk KO",
                                                     "Old, WT", "Old, MertkCR")))
  return(out) }) %>% 
  bind_rows()

pairwise_comps_aging = pairwise_comps %>% 
  left_join(., max_counts) %>% 
  dplyr::select(external_gene_name, padj, group1, group2, max_counts) %>% 
  dplyr::filter(padj < 0.05 & external_gene_name %in% aging_markers) %>% 
  group_by(external_gene_name) %>% 
  dplyr::mutate(yval = seq(from = unique(max_counts) * 1.07, 
                           by = unique(max_counts) * 0.07, 
                           length.out = n())) %>% 
  ungroup() %>% 
  dplyr::mutate(annot = format(padj, scientific = T, digits = 2),
                external_gene_name = factor(external_gene_name,
                                            levels = c("Apoe", "Cd9", "Cst7", "Lpl", "Itgax", "Spp1",
                                                        "Ccl12", "Ifit2", "Ifit3", "Ifi204", "Oasl2", "Trim30a")),
                group1 = factor(group1,
                                levels = c("Young, WT", "Young, Driver Control",
                                              "Young, Mertk KO",
                                              "Old, WT", "Old, MertkCR")),
                group2 = factor(group2,
                                levels = c("Young, WT", "Young, Driver Control",
                                              "Young, Mertk KO",
                                              "Old, WT", "Old, MertkCR")))
```

```{r}
aging_marker_boxplots = aging_marker_counts %>% 
  ggplot(aes(x = mertk_age, y = counts, fill = mertk_age)) +
  facet_wrap(~external_gene_name, scales = "free_y") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  scale_fill_npg() +
  labs(x = "", y = "Counts") +
  geom_signif(inherit.aes = F,
              data = pairwise_comps_aging,
              aes(xmin = group1, xmax = group2, annotations = annot, y_position = yval),
              tip_length = 0,
              textsize = 2,
              manual=TRUE) +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        strip.text.x = element_text(size = 24),
        axis.title.y = element_text(size = 18),
        plot.margin = margin(t = 5.5, r = 5.5, b = 5.5, l = 55, unit = "pt")) +
  scale_x_discrete(drop = FALSE) #really dumb bug!!

CairoPDF([REDACTED_FILE_PATH]
    width = 16,
    height = 14,
    family = "Arial")
aging_marker_boxplots
dev.off()
CairoPNG([REDACTED_FILE_PATH]
    width = 16,
    height = 14,
    units = "in",
    res = 300,
    family = "Arial")
aging_marker_boxplots
dev.off()

saveRDS(aging_marker_boxplots, [REDACTED_FILE_PATH]

aging_marker_boxplots
```

```{bash eval=FALSE}
cd [REDACTED_FILE_PATH]
mkdir merged_fastqs
cd merged_fastqs

prefix=/gpfs[REDACTED_FILE_PATH]
j=1
for i in {401..428}
do 
  cur_files=${prefix}${i}_S${j}*'_R1_001.fastq.gz'
  echo $cur_files
  out_file='GRSB2200'${i}_S${j}'_merged_R1_001.fastq.gz'
  echo $out_file
  cat $cur_files > $out_file
  ((j++))
done
```   

```{r}
counts(minimal, normalized = F) %>% 
  as.data.frame() %>% 
  write_csv(., [REDACTED_FILE_PATH]

filenames = read.csv([REDACTED_FILE_PATH] %>% 
 #handle concatenation
  dplyr::mutate(filename = basename(fastq_1),
                filename = gsub("L00\\d", "merged", filename)) %>% 
  dplyr::select(sample, filename) %>% 
  unique()

geo_md = colData(minimal) %>% 
  as.data.frame() %>% 
  dplyr::select(sample = mouse, age_months = age_at_takedown, genotype, mertk_age) %>% 
  dplyr::mutate(genotype = gsub(" ", "", genotype),
                mertk_age = gsub(" ", "", mertk_age),
                mertk_age = gsub(",", "_", mertk_age),
                sex = "Male") %>% 
  group_by(mertk_age) %>% 
  dplyr::mutate(title = paste(mertk_age, seq(from = 1, to = n()), sep = "_")) %>% 
  left_join(., filenames)

write_csv(geo_md, [REDACTED_FILE_PATH]

#cleaned version for pairing with counts
geo_md %>% 
  dplyr::select(-c(filename, title)) %>% 
  write_csv(., [REDACTED_FILE_PATH]
```   

```{r}
merged_fastqs = list.files([REDACTED_FILE_PATH]
                           pattern = "fastq\\.gz",
                           full.names = TRUE)
newer_fastqs = list.files([REDACTED_FILE_PATH]
                          pattern = "fastq\\.gz",
                           full.names = TRUE)
md5s = data.frame(path = c(merged_fastqs, 
                           newer_fastqs,
                           [REDACTED_FILE_PATH] 
                           [REDACTED_FILE_PATH] %>% 
dplyr::mutate(md5 = tools::md5sum(path),
              basename = basename(path)) %>% 
dplyr::select(path, basename, md5) %>% 
  dplyr::filter(basename %in% geo_md$filename |
                  grepl("\\.csv", basename))

write_csv(md5s, [REDACTED_FILE_PATH]
```

```{r}
dea_dfs = pairwise_minimal$hits
names(dea_dfs) = names(dea_dfs) %>% 
  gsub("mertk_age_", "", .) %>% 
  gsub("Young", "Y", .) %>% 
  gsub("Old", "O", .) %>% 
  gsub("_over_", " vs ", .) %>% 
  gsub("Wild-Type", "WT", .) %>% 
  gsub("Cx3cr1-CreER x zsGreen-lsl", "Driver Ctrl", .) %>% 
  gsub("Cx3cr1-CreER; zsGreen-lsl x Mertk-flox-flox", "Mertk KO", .)

openxlsx::write.xlsx(dea_dfs, [REDACTED_FILE_PATH]
```